File: app/src/main/java/org/runnerup/features/CreateAdvancedWorkout.java
package org.runnerup.features;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.io.File;
import java.io.IOException;
import org.json.JSONException;
import org.runnerup.R;
import org.runnerup.core.util.ViewUtil;
import org.runnerup.ui.common.widget.TitleSpinner;
import org.runnerup.core.workout.RepeatStep;
import org.runnerup.core.workout.Step;
import org.runnerup.core.workout.Workout;
import org.runnerup.core.workout.WorkoutSerializer;

public class CreateAdvancedWorkout extends AppCompatActivity {

  private Workout advancedWorkout = null;
  private TitleSpinner advancedWorkoutSpinner = null;
  private WorkoutEditorStepsAdapter advancedWorkoutStepsAdapter = null;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);

    setContentView(R.layout.create_advanced_workout);

    Intent intent = getIntent();
    String advWorkoutName = intent.getStringExtra(ManageWorkoutsActivity.WORKOUT_NAME);
    boolean workoutExists = intent.getBooleanExtra(ManageWorkoutsActivity.WORKOUT_EXISTS, false);

    advancedWorkoutSpinner = findViewById(R.id.new_workout_spinner);
    advancedWorkoutSpinner.setValue(advWorkoutName);
    advancedWorkoutSpinner.setEnabled(false);

    RecyclerView advancedStepList = findViewById(R.id.new_advnced_workout_steps);
    advancedStepList.setLayoutManager(new LinearLayoutManager(this));
    advancedWorkoutStepsAdapter = new WorkoutEditorStepsAdapter(this, onWorkoutChanged);
    advancedStepList.setAdapter(advancedWorkoutStepsAdapter);
    advancedWorkoutStepsAdapter.attachToRecyclerView(advancedStepList);

    Button addStepButton = findViewById(R.id.add_step_button);
    addStepButton.setOnClickListener(addStepButtonClick);

    Button addRepeatButton = findViewById(R.id.add_repeat_button);
    addRepeatButton.setOnClickListener(addRepeatStepButtonClick);

    Button saveWorkoutButton = findViewById(R.id.workout_save_button);
    saveWorkoutButton.setOnClickListener(saveWorkoutButtonClick);

    Button discardWorkoutButton = findViewById(R.id.workout_discard_button);
    discardWorkoutButton.setOnClickListener(discardWorkoutButtonClick);

    if (workoutExists) {
      discardWorkoutButton.setVisibility(View.GONE);
    }

    try {
      createAdvancedWorkout(advWorkoutName, workoutExists);
    } catch (Exception e) {
      handleWorkoutFileException(e);
    }

    if (getSupportActionBar() != null) {
      getSupportActionBar().setDisplayHomeAsUpEnabled(true);
      getSupportActionBar()
          .setBackgroundDrawable(
              androidx.core.content.ContextCompat.getDrawable(this, R.color.backgroundPrimary));
    }

    ViewUtil.Insets(findViewById(R.id.create_advanced_workout_view), true);
  }

  @Override
  public boolean onSupportNavigateUp() {
    finish();
    return true;
  }

  private void createAdvancedWorkout(String name, boolean workoutExists)
      throws JSONException, IOException {
    if (workoutExists) {
      advancedWorkout = WorkoutSerializer.readFile(getApplicationContext(), name);
    } else {
      advancedWorkout = new Workout();
      WorkoutSerializer.writeFile(getApplicationContext(), name, advancedWorkout);
    }
    refreshStepList();
  }

  private void refreshStepList() {
    advancedWorkoutStepsAdapter.setWorkout(advancedWorkout);
  }

  private final Runnable onWorkoutChanged =
      () -> {
        String advWorkoutName = advancedWorkoutSpinner.getValue().toString();
        if (advancedWorkout != null) {
          Context ctx = getApplicationContext();
          try {
            WorkoutSerializer.writeFile(ctx, advWorkoutName, advancedWorkout);
          } catch (Exception ex) {
            new AlertDialog.Builder(CreateAdvancedWorkout.this)
                .setTitle(org.runnerup.common.R.string.Failed_to_load_workout)
                .setMessage("" + ex)
                .setPositiveButton(
                    org.runnerup.common.R.string.OK, (dialog, which) -> dialog.dismiss())
                .show();
          }
        }
      };

  private final View.OnClickListener addStepButtonClick =
      v -> {
        advancedWorkout.addStep(new Step());
        refreshStepList();
      };

  private final View.OnClickListener addRepeatStepButtonClick =
      view -> {
        advancedWorkout.addStep(new RepeatStep());
        refreshStepList();
      };

  private final View.OnClickListener saveWorkoutButtonClick =
      v -> {
        try {
          String advWorkoutName = advancedWorkoutSpinner.getValue().toString();
          WorkoutSerializer.writeFile(getApplicationContext(), advWorkoutName, advancedWorkout);
          finish();
        } catch (Exception e) {
          handleWorkoutFileException(e);
        }
      };

  private void handleWorkoutFileException(Exception e) {
    new AlertDialog.Builder(CreateAdvancedWorkout.this)
        .setTitle(getString(org.runnerup.common.R.string.Failed_to_create_workout))
        .setMessage(e.toString())
        .setPositiveButton(org.runnerup.common.R.string.OK, (dialog, which) -> dialog.dismiss())
        .show();
  }

  private final View.OnClickListener discardWorkoutButtonClick =
      view ->
          new AlertDialog.Builder(CreateAdvancedWorkout.this)
              .setTitle(org.runnerup.common.R.string.Delete_workout)
              .setMessage(org.runnerup.common.R.string.Are_you_sure)
              .setPositiveButton(
                  org.runnerup.common.R.string.Yes,
                  (dialog, which) -> {
                    dialog.dismiss();
                    String name = advancedWorkoutSpinner.getValue().toString();
                    File f = WorkoutSerializer.getFile(getApplicationContext(), name);
                    //noinspection ResultOfMethodCallIgnored
                    f.delete();
                    finish();
                  })
              .setNegativeButton(
                  org.runnerup.common.R.string.No, (dialog, which) -> dialog.dismiss())
              .show();
}

File: app/src/main/kotlin/org/runnerup/features/CreateAdvancedWorkout.kt
/*
 * Copyright (C) 2013 jonas.oreland@gmail.com
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 */

package org.runnerup.features

import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.io.File
import org.json.JSONException
import org.runnerup.R
import org.runnerup.core.util.ViewUtil
import org.runnerup.core.workout.RepeatStep
import org.runnerup.core.workout.Step
import org.runnerup.core.workout.Workout
import org.runnerup.core.workout.WorkoutSerializer

class CreateAdvancedWorkout : AppCompatActivity() {

  private var advancedWorkout: Workout? = null
  private var advancedWorkoutStepsAdapter: WorkoutEditorStepsAdapter? = null
  private var workoutName: String = ""
  private var workoutExists = false

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
    setContentView(R.layout.create_advanced_workout)

    workoutName = intent.getStringExtra(ManageWorkoutsActivity.WORKOUT_NAME) ?: ""
    workoutExists = intent.getBooleanExtra(ManageWorkoutsActivity.WORKOUT_EXISTS, false)

    setSupportActionBar(findViewById(R.id.workout_editor_toolbar))
    supportActionBar?.apply {
      setDisplayHomeAsUpEnabled(true)
      title = workoutName
    }

    val advancedStepList = findViewById<RecyclerView>(R.id.new_advnced_workout_steps)
    advancedStepList.layoutManager = LinearLayoutManager(this)
    advancedWorkoutStepsAdapter = WorkoutEditorStepsAdapter(this, onWorkoutChanged)
    advancedWorkoutStepsAdapter!!.attachToRecyclerView(advancedStepList)
    advancedStepList.adapter = advancedWorkoutStepsAdapter

    try {
      createAdvancedWorkout(workoutName, workoutExists)
    } catch (e: Exception) {
      handleWorkoutFileException(e)
      return
    }

    ViewUtil.Insets(findViewById(R.id.create_advanced_workout_view), true)
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.workout_editor_menu, menu)
    return true
  }

  override fun onPrepareOptionsMenu(menu: Menu): Boolean {
    menu.findItem(R.id.menu_workout_discard)?.isVisible = !workoutExists
    return super.onPrepareOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.menu_workout_add_step -> {
        addStep()
        return true
      }
      R.id.menu_workout_add_repeat -> {
        addRepeat()
        return true
      }
      R.id.menu_workout_rename -> {
        showRenameDialog()
        return true
      }
      R.id.menu_workout_save -> {
        saveAndFinish()
        return true
      }
      R.id.menu_workout_discard -> {
        confirmDiscard()
        return true
      }
    }
    return super.onOptionsItemSelected(item)
  }

  override fun onSupportNavigateUp(): Boolean {
    finish()
    return true
  }

  @Throws(JSONException::class, java.io.IOException::class)
  private fun createAdvancedWorkout(name: String, exists: Boolean) {
    advancedWorkout =
        if (exists) {
          WorkoutSerializer.readFile(applicationContext, name)
        } else {
          Workout().also { WorkoutSerializer.writeFile(applicationContext, name, it) }
        }
    refreshStepList()
  }

  private fun refreshStepList() {
    advancedWorkout?.let { advancedWorkoutStepsAdapter?.setWorkout(it) }
  }

  private val onWorkoutChanged =
      Runnable {
        val workout = advancedWorkout ?: return@Runnable
        try {
          WorkoutSerializer.writeFile(applicationContext, workoutName, workout)
        } catch (ex: Exception) {
          AlertDialog.Builder(this@CreateAdvancedWorkout, R.style.AlertDialogTheme)
              .setTitle(org.runnerup.common.R.string.Failed_to_load_workout)
              .setMessage(ex.toString())
              .setPositiveButton(org.runnerup.common.R.string.OK) { dialog, _ -> dialog.dismiss() }
              .show()
        }
      }

  private fun addStep() {
    advancedWorkout?.addStep(Step())
    refreshStepList()
  }

  private fun addRepeat() {
    advancedWorkout?.addStep(RepeatStep())
    refreshStepList()
  }

  private fun saveAndFinish() {
    val workout = advancedWorkout ?: return
    try {
      WorkoutSerializer.writeFile(applicationContext, workoutName, workout)
      finish()
    } catch (e: Exception) {
      handleWorkoutFileException(e)
    }
  }

  private fun showRenameDialog() {
    val input =
        EditText(this).apply {
          setText(workoutName)
          setSelection(text.length)
        }
    AlertDialog.Builder(this, R.style.AlertDialogTheme)
        .setTitle(R.string.Rename_workout)
        .setMessage(org.runnerup.common.R.string.Set_workout_name)
        .setView(input)
        .setPositiveButton(org.runnerup.common.R.string.OK) { _, _ ->
          val newName = input.text.toString().trim()
          if (newName.isNotEmpty() && newName != workoutName) {
            renameWorkout(newName)
          }
        }
        .setNegativeButton(org.runnerup.common.R.string.Cancel) { dialog, _ -> dialog.dismiss() }
        .show()
  }

  private fun renameWorkout(newName: String) {
    val ctx: Context = applicationContext
    val oldFile = WorkoutSerializer.getFile(ctx, workoutName)
    val newFile = WorkoutSerializer.getFile(ctx, newName)
    if (newFile.exists()) {
      Toast.makeText(this, R.string.Workout_name_exists, Toast.LENGTH_LONG).show()
      return
    }
    if (!oldFile.renameTo(newFile)) {
      Toast.makeText(
              this,
              getString(org.runnerup.common.R.string.Failed_to_create_workout),
              Toast.LENGTH_LONG,
          )
          .show()
      return
    }
    val pref = PreferenceManager.getDefaultSharedPreferences(this)
    val prefKey = getString(R.string.pref_advanced_workout)
    if (workoutName == pref.getString(prefKey, "")) {
      pref.edit().putString(prefKey, newName).apply()
    }
    workoutName = newName
    supportActionBar?.title = workoutName
    onWorkoutChanged.run()
  }

  private fun handleWorkoutFileException(e: Exception) {
    AlertDialog.Builder(this, R.style.AlertDialogTheme)
        .setTitle(getString(org.runnerup.common.R.string.Failed_to_create_workout))
        .setMessage(e.toString())
        .setPositiveButton(org.runnerup.common.R.string.OK) { dialog, _ -> dialog.dismiss() }
        .show()
  }

  private fun confirmDiscard() {
    AlertDialog.Builder(this, R.style.AlertDialogTheme)
        .setTitle(org.runnerup.common.R.string.Delete_workout)
        .setMessage(org.runnerup.common.R.string.Are_you_sure)
        .setPositiveButton(org.runnerup.common.R.string.Yes) { dialog, _ ->
          dialog.dismiss()
          WorkoutSerializer.getFile(applicationContext, workoutName).delete()
          finish()
        }
        .setNegativeButton(org.runnerup.common.R.string.No) { dialog, _ -> dialog.dismiss() }
        .show()
  }
}
